File: Data Structures & Algorithms/house-robber-ii/submission-3.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        if len(nums) == 1:
            return nums[0]
        
        def helper(nums,start,end):
            prev = 0
            prev2 = 0

            currentmax = 0
            for i in range(start, end + 1):
                currentmax = max(prev, prev2 + nums[i])

                prev2 = prev
                prev = currentmax

            return currentmax

        return max(helper(nums,0,len(nums)-2), helper(nums,1,len(nums)-1))

File: Data Structures & Algorithms/house-robber/submission-3.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        int prev = 0;
        int prev2 = 0;
        int current_max = 0;

        for (int num: nums){
            current_max = std::max(prev, prev2 + num);

            prev2 = prev;
            prev = current_max;
        }

        return current_max;
        
    }
};
